lib/buffer.js

```diff
@@ -502,27 +561,29 @@ Buffer.prototype.write = function(string, offset, length, encoding) {
   for (;;) {
     switch (encoding) {
       case 'hex':
-        return this.hexWrite(string, offset, length);
+        if (string.length % 2 !== 0)
+          throw new TypeError('Invalid hex string');
```

Comment:
okay, maybe not all of it (like checking `instanceof Buffer`), but cases like this should.